crucible/src/Lang/Crucible/Concretize.hs

```diff
@@ -0,0 +1,380 @@
+-----------------------------------------------------------------------
+-- |
+-- Module           : Lang.Crucible.Concretize
+-- Description      : Get feasible concrete values from a model
+-- Copyright        : (c) Galois, Inc 2024
+-- License          : BSD3
+-- Maintainer       : Langston Barrett <langston@galois.com>
+-- Stability        : provisional
+--
+-- This module defines 'concRegValue', a function that takes a 'RegValue' (i.e.,
+-- a symbolic value), and a model from the SMT solver ('W4GE.GroundEvalFn'), and
+-- returns the concrete value that the symbolic value takes in the model.
+--
+-- This can be used to report specific values that lead to violations of
+-- assertions, including safety assertions.
+------------------------------------------------------------------------
+
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE StandaloneKindSignatures #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Lang.Crucible.Concretize
+  ( ConcRegValue
+  , ConcRegValue'(..)
+  , ConcAnyValue(..)
+  , ConcIntrinsic
+  , IntrinsicConcFn(..)
+  , ConcCtx(..)
+  , concRegValue
+  ) where
+
+import           Data.Kind (Type)
+import           Data.Map (Map)
+import qualified Data.Map as Map
+import           Data.Text (Text)
+import qualified Data.Vector as V
+import           Data.Word (Word16)
+
+import qualified Data.Parameterized.Context as Ctx
+import           Data.Parameterized.Map (MapF)
+import qualified Data.Parameterized.Map as MapF
+
+import           What4.Expr (Expr)
+import qualified What4.Expr.GroundEval as W4GE
+import           What4.Interface (SymExpr)
+import qualified What4.Interface as W4I
+import qualified What4.Partial as W4P
+
+import           Lang.Crucible.FunctionHandle (FnHandle, RefCell)
+import           Lang.Crucible.Simulator.Intrinsics (Intrinsic)
+import           Lang.Crucible.Simulator.RegValue (RegValue, FnVal)
+import qualified Lang.Crucible.Simulator.RegValue as RV
+import qualified Lang.Crucible.Simulator.SymSequence as SymSeq
+import qualified Lang.Crucible.Utils.MuxTree as MuxTree
+import           Lang.Crucible.Types
+
+-- | Newtype to allow partial application of 'ConcRegValue'.
+--
+-- Type families cannot appear partially applied.
+type ConcRegValue' :: Type -> CrucibleType -> Type
+newtype ConcRegValue' sym tp = ConcRegValue' { getConcRegValue' :: ConcRegValue sym tp }
+
+-- | Defines the \"concrete\" interpretations of 'CrucibleType' (as opposed
+-- to the \"symbolic\" interpretations, which are defined by 'RegValue'), as
+-- returned by 'concRegValue'.
+type ConcRegValue :: Type -> CrucibleType -> Type
+type family ConcRegValue sym tp where
```

Comment:
I am very surprised that we have to pass a `sym` type parameter here. In my mind, `ConcRegValue` should exist independently of the symbolic backend, much like `what4`'s `GroundValue` does.

Reply:
It's needed in this case:
```haskell
  ConcRegValue sym (FloatType fi) = W4GE.GroundValue (SymInterpretedFloatType sym fi)
```
I think the issue is that Crucible has "interpreted" floats, whereas What4's "base" floats are somehow different?

Reply:
A Haddock says:
```
   -- | A type index for floating point numbers, whose interpretation
   --   depends on the symbolic backend.
```
which makes it sound like this might be unavoidable. I could note the cause in the Haddock, though?

Reply:
Ah, I missed the fact that `FloatType` is parameterized by a `FloatInfo`. In the case, I agree that passing a `sym` here is likely the most direct way of making this work. (It would be nice to comment that this is the one reason why we have to pass a `sym`.)

For context: both [`SymFloat`](https://hackage.haskell.org/package/what4-1.6/docs/What4-Interface.html#t:SymFloat) and [`SymInterpretedFloat`](https://hackage.haskell.org/package/what4-1.6/docs/What4-InterpretedFloatingPoint.html#t:SymInterpretedFloat) are What4 concepts. `SymFloat` always uses an IEEE-754 interpretation of symbolic floats, whereas `SymInterpretedFloat` can use IEEE-754, real numbers, or uninterpreted functions depending on how the symbolic backend is configured. The latter is more flexible, which is why Crucible uses it.